jvm/src/main/kotlin/app/cash/trifle/Certificate.kt

```diff
@@ -48,35 +48,20 @@ data class Certificate internal constructor(
    * @param date - The date to use for verification against certificates' validity windows. If null,
    *   the current time is used.
    *
-   * @return - VerifyResult status indicating issue if certificate is not signed by the given root
-   *   certificate, or if the information present doesn't match that of the certificateRequest.
-   *   SUCCESS otherwise.
+   * @return - Boolean status indicating error in validation or success.
+   *   Exception thrown if certificate is not signed by the given root certificate, or
+   *   certificate is expired or invalid or
+   *   if the information present doesn't match that of the certificateRequest.
    */
   fun verify(
     certificateRequest: CertificateRequest,
     ancestorCertificateChain: List<Certificate>,
     anchorCertificate: Certificate,
     date: Date? = null
-  ): VerifyResult {
+  ): Boolean {
```

Comment:
Why replace the nice VerifyResult type with exception logic?

Reply:
(1) to make  the behavior of Certificate.verify and SignedData.verify consistent and (2) Certificate.verify  calls the function that throws anyway. Instead of catching it and then returning it as verifyResult, I am letting the calling app catch it

Basically, we needed a way to return error in SignedData which currently returns the SignedData object if true and null otherwise. I had a couple of options for returning error:

- changed SignedData object to have a error field with the behavior that either error is set with the reason why validate fails or signed Data fields are set
- throw an exception with the reason why validate fails

The first  could be too error prone - what if both fields are set, or neither is set, We want to return SignedData only if success.

Finally, the only other option I have is using the Kotlin https://kotlinlang.org/api/latest/jvm/stdlib/kotlin/-result/ but from what ive read for specific application errors, it's better to return exceptions.

Reply:
I'd argue that `Certificate.verify` and `SignedData.verifyAndExtract` are probably more analogous to each other. I am also not fully onboard with re-throwing exceptions in the API. This pushes the burden of the exception handling to the client to reason about which method is throwing and how to handle - especially so with nested methods that throw. What prevents a user from doing a try catch all the `Exception` type? This, however, can be represented slightly better with the use of monads such as Result helps to solve.

Reply:
I personally lean towards exceptions. I also talked to the client teams and that was their preferred approach as well. Check out this [answer](https://stackoverflow.com/questions/70847513/when-and-how-to-use-result-in-kotlin) for when to use generic result type.

Reply:
Just quoting from the article linked in the stackoverflow thread:
>>> use exceptions for logic errors, type-safe results for everything else. Don’t use exceptions as a work-around to sneak a result value out of a function.

>>> Modelling a set of outcomes can be done with a custom sealed class that cover different cases, like a result type but with specific error subtypes (and more interesting business data about the error than just Throwable).

https://adambennett.dev/2020/05/the-result-monad/